keps/sig-node/20190626-node-shutdown-state.md

```diff
@@ -0,0 +1,237 @@
+[2] [Gc orphaned pods](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/gc_controller.go#L97)
+[3] [requirements on node objects](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L167)
+[8] [bootstrap checkpointing](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/0004-bootstrap-checkpointing.md)
+[9] [bootstrap checkpointing deprecation intent](https://github.com/kubernetes/enhancements/issues/1103)
+
+### Goals
+
+* Increase the availability of stateful workloads
+* Automate self-healing for stateful workloads
+
+### Non-Goals
+
+* Introduce a new API to report node state
+* Handle cases where the Kubelet itself is unresponsive
+
+
+## Proposal
+
+user stories:
+
+* as a cluster administrator I want my stateful workload to failover in case of a node is shutdown without any intervention
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP do not intend to introduce any new API, instead we plan on re-using the Lease API and a mechanism which is similar to leader election (if not the same).
+As stated this KEP is all about finding the right inter-lock between all the actors. The main are:
+
+* Kube-Controller-Manager
+* Kubelets
+
+
+
+#### kube-controller-manager
+
+
+kube-controller-manager needs to provide a controller that:
+
+Calls the cloud provider
+Apply the node-shutdown taint
+Executes the detach operation
+Removes the taint
+
+given the current design of our controllers and their following of the single-responsibility pattern, we’ll introduce a new controller called gc_volume_controller under pkg/controller/gc, we will also make the attach_detach_controller skip shutdown nodes to avoid concurrent detach requests.
+
+This gc_volume_controller watches node objects and when notified (each step describe the possible race condition that could happen)
+
+1. checks the node state via IsShutdownByProviderID(): if there’s a race condition with the cloud provider (i.e. the instance came back see 2.a)
+2. apply a node-shutdown taint
+3. tries to hold node heartbeat lease
+  3.a. fails: backoff and retry if max retries hit -> aborts
+  3.b. success: hold the lease for this node: if the node comes back at this moment it won’t be able to acquire the lease -> pods won’t start
+4. register node as being processed in order to renew the Lease periodically
+5. delete the pod with a 0-grace-period
+6. detach the volumes
+7. once 6. returns untaint the node and stop renewing the Lease
+
+
+To avoid being too reactive in 2. we can backoff and not try to acquire the lock immediately to avoid evicting too fast
+
+#### Kubelet
+
+we need also to apply changes to the kubelet, at startup the kubelet will:
+
+1. acquire the Lease:
+  a. fails: do not start any pods, unless the pod is:
+     i. Tolerate the node shutdown taint
```

Comment:
The assumption implicit here is that "kubelet can't acquire its lease" implies that "the node is being shut down".  Otherwise, why would toleration of the shutdown taint allow the kubelet to run pods without being able to acquire its lease?  I don't think this assumption will always hold, as operators could choose to use the "acquire the node's lease" tool to prevent the kubelet from acting in other, non-shutdown scenarios.  This is another reason not to make the taint behavior-related, rather than scenario-specific.

Reply:
What happens if the kubelet start but cannot reach the apiserver ?

Reply:
Then it should assume it _could be_ in the process of being shut down, and not run pods that do not tolerate the taint.  Otherwise the bug you are fixing comes back again...

I'm not sure how that relates to the point i'm making though.

Reply:
> I don't think this assumption will always hold, as operators could choose to use the "acquire the node's lease" tool to prevent the kubelet from acting in other, non-shutdown scenarios. This is another reason not to make the taint behavior-related, rather than scenario-specific.

If operators want to do it then they're aware of the issues, also operators can always mimic what control-plane can do and use it for other non-supported scenarios